homeassistant/components/light/tellstick.py

```diff
@@ -5,126 +5,84 @@
-        for light_device in lights:
-            if light_device.tellstick_device.id == id_:
-                # Execute the update in another thread
-                light_device.update_ha_state(True)
-                break
-
-    callback_id = core.register_device_event(_device_event_callback)
-
-    def unload_telldus_lib(event):
-        """Un-register the callback bindings."""
-        if callback_id is not None:
-            core.unregister_callback(callback_id)
+    if (discovery_info is None or
+            discovery_info[DISCOVER_DEVICES] is None or
+            TELLCORE_REGISTRY is None):
+        return
 
-    hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, unload_telldus_lib)
+    signal_repetitions = discovery_info.get(DISCOVER_CONFIG,
+                                            SIGNAL_REPETITIONS)
 
-    add_devices_callback(lights)
+    add_devices(TellstickLight(
+        TELLCORE_REGISTRY.get_device(switch_id), signal_repetitions)
+                for switch_id in discovery_info[DISCOVER_DEVICES])
 
 
-class TellstickLight(Light):
+class TellstickLight(TellstickDevice, Light):
     """Representation of a Tellstick light."""
 
     def __init__(self, tellstick_device, signal_repetitions):
         """Initialize the light."""
-        import tellcore.constants as tellcore_constants
-
-        self.tellstick_device = tellstick_device
-        self.signal_repetitions = signal_repetitions
         self._brightness = 0
-
-        self.last_sent_command_mask = (tellcore_constants.TELLSTICK_TURNON |
-                                       tellcore_constants.TELLSTICK_TURNOFF |
-                                       tellcore_constants.TELLSTICK_DIM |
-                                       tellcore_constants.TELLSTICK_UP |
-                                       tellcore_constants.TELLSTICK_DOWN)
-        self.update()
-
-    @property
-    def name(self):
-        """Return the name of the switch if any."""
-        return self.tellstick_device.name
+        TellstickDevice.__init__(self, tellstick_device, signal_repetitions)
 
     @property
     def is_on(self):
         """Return true if switch is on."""
-        return self._brightness > 0
+        return self._state
 
     @property
     def brightness(self):
         """Return the brightness of this light between 0..255."""
         return self._brightness
 
-    def turn_off(self, **kwargs):
-        """Turn the switch off."""
-        for _ in range(self.signal_repetitions):
+    def set_tellstick_state(self, last_command_sent, last_data_sent):
+        """Update the internal representation of the switch."""
+        from tellcore.constants import TELLSTICK_TURNON, TELLSTICK_DIM
+        if last_command_sent == TELLSTICK_DIM:
+            if last_data_sent is not None:
+                self._brightness = int(last_data_sent)
+            self._state = self._brightness > 0
+        elif last_command_sent == TELLSTICK_TURNON:
+            self._brightness = 255
```

Comment:
You would want to set brightness to `None`. Otherwise the frontend will show a brightness slider while that is not possible.